dashboard/app/controllers/registrations_controller.rb

```diff
@@ -137,4 +141,28 @@ def update_params(params)
       races: []
     )
   end
+
+  # resource_name, resource, devise_mapping, and resource_class are defined here mostly following
+  # https://pupeno.com/2016/04/26/show-a-devise-log-in-or-sign-up-forms-in-another-page/
+  # in order to allow the upgrade action to render an error page correctly
```

Comment:
Did you try `prepend_before_action :authenticate_scope!, only: :upgrade`, as suggested in [slack](https://codedotorg.slack.com/archives/C0T10HG6N/p1499379603621506)? That method supposedly [gets the resource](https://github.com/plataformatec/devise/blob/master/app/controllers/devise/registrations_controller.rb#L131), though I haven't looked into it. It would be nice to only need that line. If that doesn't work however, then I'm fine with this approach.